core/src/Revolution/modX.php

```diff
@@ -1703,88 +1702,90 @@ public function invokeEvent($eventName, array $params= array ()) {
     /**
      * Loads and runs a specific processor.
      *
-     * @param string $action The processor to run, eg: context/update
+     * @param string $action The processor to run, eg: context/update, or the processor class name (\MODX\Revolution\Processors\Context\Update::class)
      * @param array $scriptProperties Optional. An array of parameters to pass to the processor.
      * @param array $options Optional. An array of options for running the processor, such as:
-     *
      * - processors_path - If specified, will override the default MODX processors path.
-     * - location - A prefix to load processor files from, will prepend to the action parameter
-     * (Note: location will be deprecated in future Revolution versions.)
      *
      * @return mixed The result of the processor.
      */
     public function runProcessor($action = '', $scriptProperties = array(), $options = array()) {
         $result = null;
 
-        if (class_exists($action)) {
+        // Make sure the required services are loaded before initialising a processor
+        if (!$this->lexicon) {
+            $this->services->get('lexicon');
+        }
+        if (!$this->error) {
+            $this->services->get('error');
+        }
+
+        // First check if the processor can be found directly as a class name provided to $action
+        // We're limiting to subclasses of Processor to prevent instantiating arbitrary classes
+        if (class_exists($action) && is_subclass_of($action, Processor::class, true)) {
             /** @var Processor $processor */
             $processor = new $action($this, $scriptProperties);
 
             return $processor->run();
         }
 
+        // Check if we're dealing with an action like "context/update", which we transform into the 3.0+ class name
         $legacyAction = 'MODX\\Revolution\\Processors\\' . implode('\\', array_map('ucfirst', explode('/', $action)));
+
+        // Special case for TVs, for which the directory is named differently from how it was called in the past
         if (strpos($legacyAction, '\\Tv\\') !== false) {
             $legacyAction = str_replace('\\Tv\\', '\\TemplateVar\\', $legacyAction);
         }
-        if (class_exists($legacyAction)) {
+        if (class_exists($legacyAction) && is_subclass_of($legacyAction, Processor::class, true)) {
```

Comment:
This conditional statement is being used three times in this PR, maybe it's better to make a function out of it?